src/main/daml/Daml/Finance/Instrument/Swap/Fpml/Util.daml

```diff
@@ -327,7 +329,114 @@ roundRate rate rounding =
         Down -> RoundingFloor
         Nearest -> RoundingHalfUp
 
--- | Create claims from swapStream that describes a fixed or floating coupon stream.
+-- | Create claims from swapStream that describes a floating coupon stream.
+-- This function can handle compounding of rates (currently only straight compounding)
+-- Integrate this in calculateFloatingPaymentClaimsFromSwapStream once the other compounding
+-- methods have been implemented.
+calculateCompoundedFloatingPaymentClaimsFromSwapStream : FloatingRateCalculation -> SwapStream ->
+  PeriodicSchedule -> [SchedulePeriod] -> [SchedulePeriod] -> Bool -> Bool -> Deliverable ->
+  Party -> Party -> Optional Text -> Optional [Date] -> [(Decimal, Bool)] -> Update [TaggedClaim]
+calculateCompoundedFloatingPaymentClaimsFromSwapStream floatingRateCalculation s periodicSchedule
+  calculationSchedule paymentSchedule useAdjustedDatesForDcf issuerPaysLeg currency issuer
+  calendarDataAgency fxRateId fxFixingDates notionals = do
+    assertMsg "rate rounding not yet supported" $ isNone floatingRateCalculation.finalRateRounding
+    (rateFixingDates, rateFixingCalendar) <- getRateFixingsAndCalendars s (fromSome s.resetDates)
+      calculationSchedule issuer calendarDataAgency
+
+    case fxFixingDates of
+      Some fxDates ->
+        assertMsg "fxFixingDates must match rateFixingDates" $ fxDates == rateFixingDates
+      None -> assertMsg "No fx fixing dates provided" True
+    let
+      fixLeg = False
+      calculationNotionalFixings = zip3 calculationSchedule notionals rateFixingDates
+
+      -- calculate the notional and fixing dates for a payment period
+      calculateNotionalAndFixings p =
```

Comment:
As a general comment, there are a few rather complex function that are defined "anonymously:, e.g. `foldRates`, `calculateNotionalAndFixings` that I feel could be defined as a stand-alone function (with signatures and proper docs) and then hidden so that they stay "private" to the module. I feel that would make the code easier to read.

Reply:
Thanks! I will look at this after I have implemented the other compounding methods. I have added it to the compounding issue: https://github.com/digital-asset/daml-finance/issues/517